spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,552 @@
+          .collect(Collectors.groupingBy(task -> task.file().partition()));
+
+      Map<StructLike, List<List<FileScanTask>>> fileGroupsByPartition = Maps.newHashMap();
+
+      filesByPartition.forEach((partition, tasks) -> {
+        List<List<FileScanTask>> fileGroups = toFileGroups(tasks);
+        if (fileGroups.size() > 0) {
+          fileGroupsByPartition.put(partition, fileGroups);
+        }
+      });
+
+      return fileGroupsByPartition;
+    } finally {
+      try {
+        fileScanTasks.close();
+      } catch (IOException io) {
+        LOG.error("Cannot properly close file iterable while planning for rewrite", io);
+      }
+    }
+  }
+
+  private List<List<FileScanTask>> toFileGroups(List<FileScanTask> tasks) {
+    Iterable<FileScanTask> filtered = strategy.selectFilesToRewrite(tasks);
+    Iterable<List<FileScanTask>> groupedTasks = strategy.planFileGroups(filtered);
+    return ImmutableList.copyOf(groupedTasks);
+  }
+
+  @VisibleForTesting
+  void rewriteFiles(RewriteExecutionContext ctx, FileGroup fileGroup,
+                    ConcurrentLinkedQueue<String> rewrittenIDs,
+                    ConcurrentHashMap<String, Pair<FileGroupInfo, FileGroupRewriteResult>> results) {
+
+    String groupID = fileGroup.groupID();
+    int filesInGroup = fileGroup.numFiles();
+
+    String desc = jobDesc(fileGroup, ctx);
+
+    Set<DataFile> addedFiles =
+        withJobGroupInfo(
+            newJobGroupInfo("REWRITE-DATA-FILES", desc),
+            () -> strategy.rewriteFiles(groupID, fileGroup.files()));
+
+    rewrittenIDs.offer(groupID);
+    FileGroupRewriteResult fileGroupResult = new FileGroupRewriteResult(addedFiles.size(), filesInGroup);
+
+    results.put(groupID, Pair.of(fileGroup.info(), fileGroupResult));
+  }
+
+  private void commitOrClean(Set<String> rewrittenIDs) {
+    try {
+      commitFileGroups(rewrittenIDs);
+    } catch (Exception e) {
+      LOG.error("Cannot commit groups {}, attempting to clean up written files", rewrittenIDs, e);
+      Tasks.foreach(rewrittenIDs)
+          .suppressFailureWhenFinished()
+          .run(this::abortFileGroup);
+      throw e;
+    }
+  }
+
+  private ExecutorService rewriteService() {
+    return MoreExecutors.getExitingExecutorService(
+        (ThreadPoolExecutor) Executors.newFixedThreadPool(
+            maxConcurrentFileGroupRewrites,
+            new ThreadFactoryBuilder()
+                .setNameFormat("Rewrite-Service-%d")
+                .build())
+    );
+  }
+
+  private Result doExecute(RewriteExecutionContext ctx, Stream<FileGroup> groupStream) {
+
+    ExecutorService rewriteService = rewriteService();
+
+    ConcurrentLinkedQueue<String> rewrittenIDs = new ConcurrentLinkedQueue<>();
+    ConcurrentHashMap<String, Pair<FileGroupInfo, FileGroupRewriteResult>> results = new ConcurrentHashMap<>();
+
+    Tasks.Builder<FileGroup> rewriteTaskBuilder = Tasks.foreach(groupStream)
+        .executeWith(rewriteService)
+        .stopOnFailure()
+        .noRetry()
```

Comment:
My feeling on this was that if the rewrite fails we are basically having a very serious issue since we are just trying to do a read.write operation in the underlying framework. If we want to do another attempt (and partial progress is enabled) a user can just rerun the operation and it would ignore completed binpacks. That said I don't have a strong opinion here but I would lean towards less configuration options if possible.